internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,84 @@
 package asset
 
 import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
 	"io"
 	"io/fs"
 	"path/filepath"
 
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
 	"github.com/spf13/afero"
+	"golang.org/x/sync/errgroup"
 )
 
-// UploadFunc is the function signature to upload contents to a destination.
-type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+// CacheUploader uploads local asset files.
+type CacheUploader struct {
+	// FS is the file system to use.
+	FS afero.Fs
 
-// UploadOpts contains optional configuration for uploading assets.
-type UploadOpts struct {
-	Reincludes []string   // Relative path under source to reinclude files that are excluded in the upload.
-	Excludes   []string   // Relative path under source to exclude in the upload.
-	UploadFn   UploadFunc // Custom implementation on how to upload the contents under a file. Defaults to S3UploadFn.
-	Recursive  bool       // Whether to walk recursively.
+	// Upload is the function called when uploading a file.
+	Upload func(path string, contents io.Reader) (string, error)
```

Comment:
nit: I get why we do this, it's just a bit strange to see a function as input vs. an interface. Perhaps, not in this PR but what do you think of the following suggestion 👇 

Is there a refactoring opportunity in the `upload/customresource` and `upload/asset` packages. Perhaps, something like:
```go
package upload

type Uploader interface {
   Upload(path string, contents io.Reader) (string, error)
}

type BucketUploader interface {
   Upload(bucket, path, data io.Reader) (string, error)
}

func ConvertS3(s3 BucketUploader) Uploader
```